packages/firestore/src/model/document.ts

```diff
@@ -100,85 +207,68 @@ export class Document extends MaybeDocument {
-    return (
-      `Document(${this.key}, ${this.version}, ${this.data.toString()}, ` +
-      `{hasLocalMutations: ${this.hasLocalMutations}}), ` +
-      `{hasCommittedMutations: ${this.hasCommittedMutations}})`
-    );
-  }
-
-  get hasPendingWrites(): boolean {
-    return this.hasLocalMutations || this.hasCommittedMutations;
-  }
-
-  static compareByField(field: FieldPath, d1: Document, d2: Document): number {
-    const v1 = d1.field(field);
-    const v2 = d2.field(field);
-    if (v1 !== undefined && v2 !== undefined) {
-      return v1.compareTo(v2);
-    } else {
-      return fail("Trying to compare documents on fields that don't exist");
+    switch (this.type) {
+      case DocumentType.UNKNOWN:
+        return `Document(UNKNOWN, ${this.key})`;
+      case DocumentType.CONTENTS_UNKNOWN:
+        return `Document(CONTENTS_UNKNOWN, ${this.key}, ${this.version})`;
+      case DocumentType.MISSING:
+        return `Document(MISSING, ${this.key}, ${this.version})`;
+      case DocumentType.EXISTS:
+        return (
+          `Document(${this.key}, ${this.version}, ${this.data.toString()}, ` +
+          `documentState=${this.documentState})`
+        );
+      default:
+        return fail('Unknown document type');
     }
   }
-}
-
-/**
- * A class representing a deleted document.
- * Version is set to 0 if we don't point to any specific time, otherwise it
- * denotes time we know it didn't exist at.
- */
-export class NoDocument extends MaybeDocument {
-  readonly hasCommittedMutations: boolean;
 
-  constructor(
-    key: DocumentKey,
-    version: SnapshotVersion,
-    options?: DocumentOptions
-  ) {
-    super(key, version);
-    this.hasCommittedMutations = !!(options && options.hasCommittedMutations);
+  get exists(): boolean {
+    return this.type === DocumentType.EXISTS;
   }
 
-  toString(): string {
-    return `NoDocument(${this.key}, ${this.version})`;
+  get missing(): boolean {
+    return this.type === DocumentType.MISSING;
   }
 
-  get hasPendingWrites(): boolean {
-    return this.hasCommittedMutations;
+  get unknown(): boolean {
+    return this.type === DocumentType.UNKNOWN;
   }
 
-  isEqual(other: MaybeDocument | null | undefined): boolean {
+  /**
+   * Returns true if the document is of some definite type: either existing
+   * or definitely missing.
+   */
+  get definite(): boolean {
     return (
-      other instanceof NoDocument &&
-      other.hasCommittedMutations === this.hasCommittedMutations &&
-      other.version.isEqual(this.version) &&
-      other.key.isEqual(this.key)
+      this.type === DocumentType.EXISTS || this.type === DocumentType.MISSING
     );
   }
-}
 
-/**
- * A class representing an existing document whose data is unknown (e.g. a
- * document that was updated without a known base document).
- */
-export class UnknownDocument extends MaybeDocument {
-  constructor(key: DocumentKey, version: SnapshotVersion) {
-    super(key, version);
+  get hasLocalMutations(): boolean {
```

Comment:
If possible I think I'd like hasLocalMutations to throw for unknown / missing documents and hasCommittedMutations to throw for unknown documents...  We can loosen if/when we need to, but ensuring code isn't checking for impossible things would be nice.